### ISSUE TITLE
Alternator: authentication: do not keep secret keys on Scylla nodes

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#5206]
Our current implementation of the DynamoDB API authentication requires that the coordinator authenticating a request know the user's **secret key**. The coordinator uses this secret key to repeat the same message signing protocol that the client does, and compares the results to verify the message's authenticity and its source. This implementation means that Scylla nodes know the secret keys. This suffers from a well-known security problem: If an attacker compromises any Scylla node at one point in time, he can fetch these secret keys, and then use these keys **in the future**, even long after the original security hole has been plugged, to sign new requests.

Amazon's signing protocol, described in https://docs.aws.amazon.com/general/latest/gr/sigv4-calculate-signature.html, actually comes built-in with a solution to this "future secrecy" problem - and obviously this was designed into the protocol and did not come about by accident. Note how the protocol does not directly use the user's secret key to HMAC the data to be authenticated. Instead it has two steps: First, a "signing key" is derived from the user's secret key. Then, it is this "signing key" which is used as the HMAC key for individual requests.

The reason why this extra "signing-key derivation" step adds security is:

1. The derivation of the "signing key" can be done on a different node, not the coordinator, possibly much better protected. Only this node will know the actual secret keys. The coordinator will never see the real secret keys - just the "signing keys" derived from them via an irreversible hash function.

2. The "signing keys" used by the coordinator can still be compromised. But now note that the calculation of the signing key includes the date (in whole-day resolution) of the request. This means that any compromised signing key will only be valid for a single day. The attacker will not be able to fake messages a week from now if the original security hole is plugged. And because the calculation of the signing key from the secret key was done in an irreversible manner (using an HMAC too), knowing today's signing key will not help the attacker find tomorrow's signing key. 

3.  Morever, the one-day resolution of the signing key means that the separate service that calculates signing keys can actually be done offline for even better security:  A stand-alone node holds the actual secret keys, and once each day calculates the "signing key" of the day for each of these keys, and writes these signing keys to a file, to be later shared with Scylla. This stand-alone signing server will be very hard to compromise, because it never accepts any user inputs and hardly interacts with anything (except writing that single file), so its attack surface is extremely limited.

**Comment by nyh** [comment]
Let's assume for a moment that we won't choose your option 1 - choosing it means closing this issue ;-)

I agree that using an Scylla table is more logical for storing the signing keys than the "file" I mentioned. After all, if you already have a hammer everything looks like nails ;-) So if we already have a database, we can store data in the database and you don't need "files".

So I propose the following plan:

1. Have a Scylla table for daily signing keys. The partition key is _user_, the clustering key is _day_ (e.g., integer number of days since the epoch). The regular column is that user's and day's daily signing key.
2. Put a TTL of (say) 25 hours on the signing key table, to get rid of no-longer-interesting old keys.
2. To _verify_ a signature, Scylla will take the timestamp written in the request's signature, read the daily signing key for _that_ day and use it to check the signature. Note that DynamoDB allows the request's time to be "off" by only a limited amount of time (IIRC, 15 minutes?) so the user can't just say this is a request from yesterday and use yesterday's key. But we do need some slack around midnight - the client's clock may think it wasn't midnight yet, so the server should respect this believe, at least for a short amount of time after midnight).
3. Where do these daily signing keys come from? Some offline operator/server/node will calculate them every day before midnight (e.g., around 23:00) for the next day, and write them to Scylla's table. The point is that when we reach midnight, Scylla should have all this data available and ready to use.
5. The signing-keys table should only be readable and writable by a one role - you don't want every user to be able to read or write other user's keys, even if those are only valid for one day. So it's similar to the role table that we already have.

**Comment by eliransin** [comment]
I agree with one addition, when I said it should be configurable, I meant that the "entity" that calculates the daily keys should be configurable. This entity can be one of:
1. Scylla (as we do today) maybe with the slight change of recalculating the key if a key for some day is missing and not every single time.
2. Some remote entity - In which case this will probably be one of two methods:
    1. Push - this is exactly what you described, some server that every interval (daily) will write the new daily keys, using
    predefined credentials 
    2. Pull - Instead of having ScyllaDB recalculating the keys, it will pull them from some service with some predefined protocol (the principal is somewhat similar to AWS KMS). This sounds very limiting but it actually isn't because you can just have a local thin server (on the loopback interface or in the same VPC or whatever) that do the adaptation for you,
    i.e communicate with some key server to retrieve this day's key.

This will allow us to have the default method as 1, and that way we will not create a regression at least. I admit I stole this modular approach from @elcallio encryption at rest where you have a basic concept of a key provider and this key provider just need to follow some basic guidelines in order to serve keys for ScyllaDB. The we go ahead and implement whatever specific key provider we chose :slightly_smiling_face:

**Comment by nyh** [comment]
>     1. Scylla (as we do today) maybe with the slight change of recalculating the key if a key for some day is missing and not every single time.

You mean that we should have a configuration option to keep the current situation. I agree that there is no harm in having such an option (which may even be the default option in some setups).

What is **not** possible is to have an option that "sometimes" Scylla calculates a missing (why?) signing key and sometimes doesn't. ‎The point is that to be able to calculate a signing key Scylla, Scylla would need to access the secret key, and this is exactly the security hole we want to avoid in this issue.

>     2. Some remote entity - In which case this will probably be one of two methods:
>        
>        1. Push - this is exactly what you described, some server that every interval (daily) will write the new daily keys, using
>           predefined credentials
>        2. Pull - Instead of having ScyllaDB recalculating the keys, it will pull them from some service with some predefined protocol (the principal is somewhat similar to AWS KMS). This sounds very limiting but it actually isn't because you can just have a local thin server (on the loopback interface or in the same VPC or whatever) that do the adaptation for you,
>           i.e communicate with some key server to retrieve this day's key.

This is indeed a possibility - Scylla can ask for a signing key when it needs it, and store the signing key in memory, as a cache. However, this approach complicates the the remote node - instead of a 100-line semi-off-line script, it becomes a full-blown server, one that needs to store its data somewhere and that needs to be highly available (whereas the "push" method I described only needs to work once a day, and it doesn't really matter when it works during the day if it gets its job done some time before midnight).

The server ("pull") approach makes a lot of sense when your company has a lot of products that want to use the same keys - as is the case in Amazon. This is not exactly the situation in ScyllaDB. The pull approach also makes sense if you have some pre-existing server you can use. I don't think that AWS signature v4 is common enough that such servers exist and we can easily re-use.
